python/ray/serve/backend_state.py

```diff
@@ -596,40 +621,70 @@ def delete_backend(self, backend_tag: BackendTag,
+
+        TODO
+        """
         # NOTE(edoakes): this short-circuits when using the legacy
         # `create_backend` codepath -- it can be removed once we deprecate
         # that as the version should never be None.
-        if version is None:
-            return
-
-        # TODO(edoakes): to implement rolling upgrade, all we should need to
-        # do is cap the number of old version replicas that are stopped here.
-        replicas_to_stop = self._replicas[backend_tag].pop(
-            exclude_version=version,
+        if target_version is None:
+            return 0
+
+        # We include SHOULD_START and STARTING replicas here because if there
+        # are replicas still pending startup, we may as well terminate them
+        # and start new version replicas instead.
+        old_running_replicas = replicas.count(
+            exclude_version=target_version,
             states=[
                 ReplicaState.SHOULD_START, ReplicaState.STARTING,
                 ReplicaState.RUNNING
             ])
-
-        # Inform the routers and backend replicas about config changes.
-        # TODO(edoakes): this should only happen if we change something other
-        # than num_replicas.
-        self._notify_backend_configs_changed(backend_tag)
-        if len(replicas_to_stop) > 0:
-            logger.info(f"Stopping {len(replicas_to_stop)} replicas of "
-                        f"backend '{backend_tag}' with outdated versions.")
+        old_stopping_replicas = replicas.count(
+            exclude_version=target_version,
+            states=[ReplicaState.SHOULD_STOP, ReplicaState.STOPPING])
+        new_running_replicas = replicas.count(
+            version=target_version, states=[ReplicaState.RUNNING])
+
+        # If the backend is currently scaling down, let the scale down
+        # complete before doing a rolling update.
+        if target_replicas < old_running_replicas + old_stopping_replicas:
+            return 0
+
+        # The number of replicas that are currently in transition between
+        # an old version and the new version. Note that we cannot directly
+        # count the number of stopping replicas because once replicas finish
+        # stopping, they are removed from the data structure.
+        pending_replicas = (
+            target_replicas - new_running_replicas - old_running_replicas)
+
+        # Maximum number of replicas that can be updating at any given time.
+        # There should never be more than rollout_size old replicas stopping
+        # or rollout_size new replicas starting.
+        rollout_size = max(int(0.2 * target_replicas), 1)
+        max_to_stop = max(rollout_size - pending_replicas, 0)
+
+        replicas_to_stop = replicas.pop(
+            exclude_version=target_version,
+            states=[
+                ReplicaState.SHOULD_START, ReplicaState.STARTING,
+                ReplicaState.RUNNING
+            ],
+            max_replicas=max_to_stop)
 
```

Comment:
in this version, seems like we are not counting number of new version replicas that are starting? namely if the following is > 0, we are breaking the invariant the max number of replicas in transition.
```
replicas.count(
            version=target_version, states=[ReplicaState.STARTING, ReplicaState.SHOULD_START])
```

Reply:
I think this part might be okay--the expression you've written is contributes to pending_replicas.

Concretely, if `target_replicas = current_replicas =` 10 replicas, and it consists of 8 old RUNNING and 2 new STARTING, then we have 

`pending_replicas = 10 - 8 - 0 = 2`

so `max_to_stop` is `2-2 = 0` and nothing gets stopped.  (Until the new_version replicas move from STARTING to RUNNING, as desired.)